File: js/stack2.js
//Stack에 대한 ADT(추상자료형)
//Array : stack을 담을 배열이자 stack 그 자체
//push : stack에 새 요소 추가
//pop : stack의 맨 위에 있는 요소 삭제
//peek : stack의 맨 위에 있는 요소 확인
//lefts: stack에 있는 모든 요소 문자열로 반환
//clear: stack에 있는 모든 요소 삭제
//empty : stack에 남은 요소가 있는지 없는지 확인

class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.top = null;
    this.bottom = null;
    this.length = 0;
  }
  peek() {
    return this.top;
  }
  push(value) {
    const newNode = new Node(value);

    if (this.length === 0) {
      this.top = newNode;
      this.bottom = newNode;
    } else {
      const holdingPointer = this.top;
      this.top = newNode;
      this.top.next = holdingPointer;
    }
    this.length++;
    return this;
  }
  pop() {
    if (!this.top) return null;
    if (this.top === this.bottom) {
      this.bottom = null;
    }
    this.top = this.top.next;
    this.length--;
    return this;
  }
}

const myStack = new Stack();
console.log(myStack);
myStack.peek();
myStack.push("google");
myStack.push("facebook");
myStack.push("udemy");
console.log(myStack.pop());
console.log(myStack.pop());
console.log(myStack.pop());

File: python/stack2.py
# python singly linked list(연결 리스트)로 스택 구현하기
# 연결 리스트의 각 원소는 노드라고한다.
# 각 노드는 데이터와 그 다음 노드를 가리키는 포인터를 가지고 있다.
class  Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedListStack():
    def __init__(self):
        self.head = None

    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def pop(self):
        pop_object = None
        if self.isEmpty():
            print('Stack is Empty')
        else:
            pop_object = self.head.data
            self.head = self.head.next
        return pop_object
    
    def top(self):
        top_object = None
        if self.isEmpty():
            print('Stack is Empty')
        else:
            top_object = self.head.data
        return top_object

    def isEmpty(self):
        is_empty = False
        if self.head is None:
            is_emtpy = True
        return is_empty